### PR TITLE
Add compilation path for lowering linalg on tensors op without using tile and distribute.

### PR DESCRIPTION
**Pull request by MaheshRavishankar**
Not all operations in a dispatch region can be lowered through tile
and distribute. Operations like `linalg.tensor_reshape` cannot be
tiled. These needs a separate lowering mechanism. In this PR, such a
mechanism is added where such operations are put into the dispatch
region as is. Its upto the individual backends to handle these
appropriately.
- On the GPU side trivially parallel operations are distribute across
  all invocations.
- On the CPU side such operations are executed sequentially for
  now. This could be modified to execute in parallel too, but the
  sequential execution is consistent with existing codegen path.

The PR also adds support for lowering linalg.tensor_reshape operation
by making LinalgBufferize handle this operation (further clean up
related to Issue #4759 with removal of dead code)

Instead of relying on ad-hoc pattern matching of the code within the
dispatch region, and explicit flag is used to distinguish between the
Linalg on tensors path and legacy path.
- On CPU the LinalgTileAndDistribute pass is split to have a pass
  MaterializeCPULaunchConfigurationPass. The former now is only for
  the legacy path, and the latter for the linalg on tensors path.
- On GPU the ConvertToGPUPass is used to distribute untiled operations
  in dispatch region to invocations.

### CODE REVIEW

**Comment by MaheshRavishankar** [comment]
PR is ready for review. Have some errors after rebasing on ToT and also have an upstream patch that I need to send out that this depends on. Still wanted to send it out for review cause its quite large PR. (apologies for that. A lot of it is refactoring and removing some ad-hoc assumptions.

File: iree/compiler/Dialect/Flow/Transforms/DispatchLinalgOnTensors.cpp [hunk 1]
```diff
@@ -302,39 +339,106 @@ struct TileAndDistributeOnTensorsPattern
...
+      continue;
+    }
+    return llvm::None;
   }
-  if (auto constantOp = dyn_cast<ConstantOp>(op)) {
-    return constantOp.getResult().getType().isIntOrFloat();
+  return outputShapes;
+}
+
+/// Rewrite pattern to put ops that are not-tilable or arent tiled. For example
```
Comment (antiagainst) [inline iree/compiler/Dialect/Flow/Transforms/DispatchLinalgOnTensors.cpp:378]: Puts stand-alone ops that are not tileable/tiled into its own `flow.dispatch_workgroups` op?

**Review by antiagainst — APPROVED** [review]

File: iree/compiler/Conversion/Common/Passes.h [hunk 2]
```diff
@@ -20,12 +20,13 @@ namespace mlir {
 namespace iree_compiler {
 
 /// Alias for callback function that allocates workgroup level memory. The
-/// callback expects the type of the memref to be allocated (`allocationType`)
-/// and the SSA values that represent the size of dynamic dimensions in the
-/// `allocationType`. The callback is expected to return a MemRefType Value.
+/// callback expects a Sha ShapedType for which the memref is being allocated.of
```
[the remark this answers is not in the document]
Comment (MaheshRavishankar) [reply iree/compiler/Conversion/Common/Passes.h:23]: oof! Thanks fixed.

File: iree/compiler/Conversion/Common/Transforms.h [hunk 3]
```diff
@@ -57,6 +57,17 @@ LogicalResult getLinalgOps(FuncOp funcOp,
                            SmallVectorImpl<linalg::LinalgOp> &linalgOps,
                            SmallVectorImpl<Operation *> &tiledLoops);
 
+/// Specifies the number of workgroups to use for a particular entry point
+/// functio, by updating the `worgroup_count` region in the
+/// `hal.executable.entry_point` op for this operation. The method takes a
+/// callback function, which computes the workgroup count (x,y, z) given the
+/// workload along (x,y,z).
+using WorkgroupCountRegionBuilder = std::function<std::array<Value, 3>(
+    OpBuilder &b, Location loc, std::array<Value, 3> workload)>;
```
[the remark this answers is not in the document]
Comment (MaheshRavishankar) [reply iree/compiler/Conversion/Common/Transforms.h:66]: Ah, I missed this in my latest commit. WIll address this.

**Pull request merged**